File: src/modules/chat-starter/lib/starter.js
// Pure chat-starter rules — no discord.js, no timers. Decides WHEN a quiet
// channel earns an ice-breaker and WHICH question to post.

export const DEFAULT_STARTER_CONFIG = {
  // Owner decision 2026-07-24: the chat starter runs in this channel after
  // 12 hours of silence — committed as product defaults (like the memorial
  // feeds) so it works out of the box; /chat-starter-config overrides win.
  enabled: true,
  channelId: '411609312037961729',
  idleMinutes: 720, // 12 hours of silence before a starter
  useAi: false, // when true AND the detective has a provider, generate instead of picking
};

export const RECENT_MEMORY = 10; // don't repeat any of the last N questions

/** Validate a question bank document. */
export function validateQuestions(doc) {
  if (!doc || !Array.isArray(doc.questions) || doc.questions.length === 0) {
    return { ok: false, error: 'questions missing or empty' };
  }
  if (!doc.questions.every((q) => typeof q === 'string' && q.trim().length > 0)) {
    return { ok: false, error: 'every question must be a non-empty string' };
  }
  return { ok: true };
}

/**
 * Should the sweep post a starter right now?
 * `humanSinceStarter` guards against the bot monologuing: after one starter,
 * at least one human message must appear before the next one.
 * @returns {{ post:true } | { post:false, reason:string }}
 */
export function shouldPost({ config, idleMs, humanSinceStarter }) {
  if (!config.enabled) return { post: false, reason: 'disabled' };
  if (!config.channelId) return { post: false, reason: 'no-channel' };
  if (!humanSinceStarter) return { post: false, reason: 'no-human-since-last-starter' };
  if (idleMs < config.idleMinutes * 60_000) return { post: false, reason: 'not-idle-enough' };
  return { post: true };
}

/**
 * Pick a question index, avoiding the recently used ones (up to RECENT_MEMORY,
 * but never excluding the entire bank). `random` injectable for tests.
 */
export function pickQuestionIndex(count, recentIndexes = [], random = Math.random) {
  const recent = new Set(recentIndexes.slice(-Math.min(RECENT_MEMORY, Math.max(0, count - 1))));
  const candidates = [];
  for (let i = 0; i < count; i += 1) if (!recent.has(i)) candidates.push(i);
  const pool = candidates.length > 0 ? candidates : [...Array(count).keys()];
  return pool[Math.floor(random() * pool.length)];
}

/** Track a used index in the recent ring. */
export function rememberIndex(recentIndexes, index) {
  return [...(recentIndexes ?? []), index].slice(-RECENT_MEMORY);
}

File: cuffchatstarter/starter.py
"""Pure chat-starter rules — port of the Node module's ``lib/starter.js``.

No discord objects and no timers, so every rule below is testable with plain
numbers. Two decisions live here: *when* a quiet channel has earned an
ice-breaker, and *which* question to post.
"""

from __future__ import annotations

import random as _random
from typing import Any, Dict, List, Optional, Sequence, Tuple

#: Owner decision 2026-07-24, carried over from the Node module: the starter
#: runs in this channel after 12 hours of silence. Committed as product
#: defaults so the cog works the moment it is loaded; commands override them.
DEFAULT_CHANNEL_ID = 411609312037961729
DEFAULT_IDLE_MINUTES = 720  # 12 hours of silence

#: Don't repeat any of the last N questions.
RECENT_MEMORY = 10

IDLE_MIN = 15
IDLE_MAX = 1440


def validate_questions(doc: Any) -> Tuple[bool, str]:
    """Check a question-bank document. Returns ``(ok, error)``."""
    if not isinstance(doc, dict):
        return False, "the bank must be a JSON object"
    questions = doc.get("questions")
    if not isinstance(questions, list) or not questions:
        return False, "questions missing or empty"
    if not all(isinstance(q, str) and q.strip() for q in questions):
        return False, "every question must be a non-empty string"
    return True, ""


def should_post(
    *, enabled: bool, channel_id: Optional[int], idle_minutes: int, idle_ms: float,
    human_since_starter: bool,
) -> Tuple[bool, str]:
    """Should the sweep post a starter right now? ``(post, reason)``.

    ``human_since_starter`` is the anti-monologue guard: after one starter, at
    least one human message must appear before the next one is allowed.
    """
    if not enabled:
        return False, "disabled"
    if not channel_id:
        return False, "no-channel"
    if not human_since_starter:
        return False, "no-human-since-last-starter"
    if idle_ms < idle_minutes * 60_000:
        return False, "not-idle-enough"
    return True, ""


def pick_question_index(
    count: int, recent_indexes: Sequence[int] = (), rng: Optional[_random.Random] = None
) -> int:
    """An index into the bank, avoiding the recently used ones.

    The recent window is clamped to ``count - 1`` so it can never exclude the
    entire bank — a 5-question bank with a memory of 10 would otherwise have
    nothing left to choose from.
    """
    if count <= 0:
        raise ValueError("empty question bank")
    rng = rng or _random
    window = min(RECENT_MEMORY, max(0, count - 1))
    recent = set(list(recent_indexes)[-window:]) if window else set()
    candidates = [i for i in range(count) if i not in recent]
    return rng.choice(candidates or list(range(count)))


def remember_index(recent_indexes: Optional[Sequence[int]], index: int) -> List[int]:
    """Track a used index in the recent ring."""
    return [*(recent_indexes or []), index][-RECENT_MEMORY:]


def new_activity(now_ms: float) -> Dict[str, Any]:
    """A fresh activity record: treat now as the last activity, starter allowed."""
    return {"last_activity_at": now_ms, "human_since_starter": True}
